lib/block_view_stake.go

```diff
@@ -0,0 +1,2496 @@
+	if err != nil {
+		return nil, err
+	}
+	if stakeEntry != nil {
+		// Cache the StakeEntry in the UtxoView if exists.
+		bav._setStakeEntryMappings(stakeEntry)
+	}
+	return stakeEntry, nil
+}
+
+func (bav *UtxoView) GetLockedStakeEntry(
+	validatorPKID *PKID,
+	stakerPKID *PKID,
+	lockedAtEpochNumber uint64,
+) (*LockedStakeEntry, error) {
+	// First, check the UtxoView.
+	lockedStakeMapKey := LockedStakeMapKey{
+		ValidatorPKID:       *validatorPKID,
+		StakerPKID:          *stakerPKID,
+		LockedAtEpochNumber: lockedAtEpochNumber,
+	}
+	if lockedStakeEntry, exists := bav.LockedStakeMapKeyToLockedStakeEntry[lockedStakeMapKey]; exists {
+		// If LockedStakeEntry.isDeleted, return nil.
+		if lockedStakeEntry.isDeleted {
+			return nil, nil
+		}
+		return lockedStakeEntry, nil
+	}
+	// Then, check the database.
+	lockedStakeEntry, err := DBGetLockedStakeEntry(bav.Handle, bav.Snapshot, validatorPKID, stakerPKID, lockedAtEpochNumber)
+	if err != nil {
+		return nil, err
+	}
+	if lockedStakeEntry != nil {
+		// Cache the LockedStakeEntry in the UtxoView if exists.
+		bav._setLockedStakeEntryMappings(lockedStakeEntry)
+	}
+	return lockedStakeEntry, nil
+}
+
+func (bav *UtxoView) GetLockedStakeEntriesInRange(
+	validatorPKID *PKID,
+	stakerPKID *PKID,
+	startEpochNumber uint64,
+	endEpochNumber uint64,
+) ([]*LockedStakeEntry, error) {
+	// Store matching LockedStakeEntries in a set to prevent
+	// returning duplicates between the db and UtxoView.
+	lockedStakeEntriesSet := NewSet([]*LockedStakeEntry{})
+
+	// First, pull matching LockedStakeEntries from the db.
+	dbLockedStakeEntries, err := DBGetLockedStakeEntriesInRange(
+		bav.Handle, bav.Snapshot, validatorPKID, stakerPKID, startEpochNumber, endEpochNumber,
+	)
+	if err != nil {
+		return nil, errors.Wrapf(err, "UtxoView.GetLockedStakeEntriesInRange: ")
+	}
+	for _, lockedStakeEntry := range dbLockedStakeEntries {
+		lockedStakeEntriesSet.Add(lockedStakeEntry)
+	}
+
+	// Then, pull matching LockedStakeEntries from the UtxoView.
+	// Loop through all LockedStakeEntries in the UtxoView.
+	for _, lockedStakeEntry := range bav.LockedStakeMapKeyToLockedStakeEntry {
+		// Filter to matching LockedStakeEntries.
+		if !lockedStakeEntry.ValidatorPKID.Eq(validatorPKID) ||
+			!lockedStakeEntry.StakerPKID.Eq(stakerPKID) ||
+			lockedStakeEntry.LockedAtEpochNumber < startEpochNumber ||
+			lockedStakeEntry.LockedAtEpochNumber > endEpochNumber {
+			continue
+		}
+
+		if lockedStakeEntry.isDeleted {
+			// Remove from set if isDeleted.
+			lockedStakeEntriesSet.Remove(lockedStakeEntry)
+		} else {
+			// Otherwise, add to set.
+			lockedStakeEntriesSet.Add(lockedStakeEntry)
+		}
+	}
+
+	// Convert LockedStakeEntries set to slice, sorted by LockedAtEpochNumber ASC.
+	lockedStakeEntries := lockedStakeEntriesSet.ToSlice()
+	sort.Slice(lockedStakeEntries, func(ii, jj int) bool {
+		return lockedStakeEntries[ii].LockedAtEpochNumber < lockedStakeEntries[jj].LockedAtEpochNumber
+	})
+	return lockedStakeEntries, nil
```

Comment:
```suggestion
	// First, pull matching LockedStakeEntries from the db.
	dbLockedStakeEntries, err := DBGetLockedStakeEntriesInRange(
		bav.Handle, bav.Snapshot, validatorPKID, stakerPKID, startEpochNumber, endEpochNumber,
	)
	if err != nil {
		return nil, errors.Wrapf(err, "UtxoView.GetLockedStakeEntriesInRange: ")
	}
	lockedStakeEntryMap := make(map[LockedStakeMapKey]*LockedStakeEntry)
	for _, lockedStakeEntry := range dbLockedStakeEntries {
		lockedStakeEntryMap[lockedStakeEntry.ToMapKey()] = lockedStakeEntry
	}

	// Then, pull matching LockedStakeEntries from the UtxoView.
	// Loop through all LockedStakeEntries in the UtxoView.
	for _, lockedStakeEntry := range bav.LockedStakeMapKeyToLockedStakeEntry {
		// Filter to matching LockedStakeEntries.
		if !lockedStakeEntry.ValidatorPKID.Eq(validatorPKID) ||
			!lockedStakeEntry.StakerPKID.Eq(stakerPKID) ||
			lockedStakeEntry.LockedAtEpochNumber < startEpochNumber ||
			lockedStakeEntry.LockedAtEpochNumber > endEpochNumber {
			continue
		}

		if lockedStakeEntry.isDeleted {
			// Remove from map if isDeleted.
			delete(lockedStakeEntryMap, lockedStakeEntry.ToMapKey())
		} else {
			// Otherwise, add to map.
			lockedStakeEntryMap[lockedStakeEntry.ToMapKey()] = lockedStakeEntry
		}
	}

	var lockedStakeEntries []*LockedStakeEntry
	// Convert LockedStakeEntries map to slice.
	for _, lockedStakeEntry := range lockedStakeEntryMap {
		lockedStakeEntries = append(lockedStakeEntries, lockedStakeEntry)
	}

	// Sort LockedStakeEntries by LockedAtEpochNumber ASC.
	sort.Slice(lockedStakeEntries, func(ii, jj int) bool {
		return lockedStakeEntries[ii].LockedAtEpochNumber < lockedStakeEntries[jj].LockedAtEpochNumber
	})
	return lockedStakeEntries, nil
```